lib/Conversion/FIRRTLToRTL/LowerToRTL.cpp

```diff
@@ -104,6 +106,78 @@ static bool isZeroBitFIRRTLType(Type type) {
   return type.cast<FIRRTLType>().getPassiveType().getBitWidthOrSentinel() == 0;
 }
 
+namespace {
+struct FirMemory {
+  size_t numReadPorts;
+  size_t numWritePorts;
+  size_t numReadWritePorts;
+  size_t dataWidth;
+  size_t depth;
+  size_t readLatency;
+  size_t writeLatency;
+  bool operator<(const FirMemory &rhs) const {
+#define cmp3way(name)                                                          \
+  if (name < rhs.name)                                                         \
+    return true;                                                               \
+  if (name > rhs.name)                                                         \
+    return false;
+    cmp3way(numReadPorts);
+    cmp3way(numWritePorts);
+    cmp3way(numReadWritePorts);
+    cmp3way(dataWidth);
+    cmp3way(depth);
+    cmp3way(readLatency);
+    cmp3way(writeLatency);
+    return false;
+#undef cmp3way
+  }
+};
+} // namespace
+
+static std::string getFirMemoryName(const FirMemory &mem) {
+  return llvm::formatv("FIRRTLMem_{0}_{1}_{2}_{3}_{4}_{5}_{6}",
+                       mem.numReadPorts, mem.numWritePorts,
+                       mem.numReadWritePorts, mem.dataWidth, mem.depth,
+                       mem.readLatency, mem.writeLatency);
+}
+
+static FirMemory analyzeMemOp(MemOp op) {
+  size_t numReadPorts = 0;
+  size_t numWritePorts = 0;
+  size_t numReadWritePorts = 0;
+
+  for (size_t i = 0, e = op.getNumResults(); i != e; ++i) {
+    auto portKind = *op.getPortKind(i);
+    if (portKind == MemOp::PortKind::Read)
+      ++numReadPorts;
+    else if (portKind == MemOp::PortKind::Write)
+      ++numReadPorts;
+    else
+      ++numReadWritePorts;
+  }
+  auto width = op.getDataType().getBitWidthOrSentinel();
+  if (width <= 0) {
+    op.emitError("'firrtl.mem' should have simple type and known width");
+    width = 0;
+  }
+  return {numReadPorts, numWritePorts,    numReadWritePorts, (size_t)width,
+          op.depth(),   op.readLatency(), op.writeLatency()};
+}
+
+static std::set<FirMemory> collectFIRRTLMemories(Operation *op) {
+  auto module = cast<FModuleOp>(op);
+  std::set<FirMemory> retval;
+  for (auto op : module.getBody().getOps<MemOp>())
+    retval.insert(analyzeMemOp(op));
+  return retval;
+}
+
+static std::set<FirMemory> mergeFIRRTLMemories(std::set<FirMemory> a,
+                                               std::set<FirMemory> b) {
+  a.insert(b.begin(), b.end());
+  return a;
+}
```

Comment:
Is this returning a copy of `a`? Can this be done more efficiently, e.g., returning a reference to the mutated `a`?